src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2164,20 +2164,107 @@ Node* OrVNode::Identity(PhaseGVN* phase) {
   return redundant_logical_identity(this);
 }
 
+// Returns whether (XorV (VectorMaskCmp) -1) can be optimized by performing the
+// inverse of a comparison operation.
+bool VectorMaskCmpNode::predicate_can_be_inverted() {
+  switch (_predicate) {
+    case BoolTest::eq:
+    case BoolTest::ne:
+      // eq and ne also apply to floating-point special values like NaN and infinities.
+      return true;
+    case BoolTest::le:
+    case BoolTest::ge:
+    case BoolTest::lt:
+    case BoolTest::gt:
+    case BoolTest::ule:
+    case BoolTest::uge:
+    case BoolTest::ult:
+    case BoolTest::ugt: {
+      BasicType bt = vect_type()->element_basic_type();
+      // For float and double, we don't know if either comparison operand is a
+      // NaN, NaN {le|ge|lt|gt} anything is false, resulting in inconsistent
+      // results before and after negation.
+      return is_integral_type(bt);
+    }
+    default:
+      return false;
+  }
+}
+
 Node* XorVNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  Node* in1 = in(1);
+  Node* in2 = in(2);
   // (XorV src src)      => (Replicate zero)
   // (XorVMask src src)  => (MaskAll zero)
   //
   // The transformation is only applied to the un-predicated
   // operation, since the VectorAPI masked operation requires
   // the unmasked lanes to save the same values in the first
   // operand.
-  if (!is_predicated_vector() && (in(1) == in(2))) {
+  if (!is_predicated_vector() && (in1 == in2)) {
     BasicType bt = vect_type()->element_basic_type();
     Node* zero = phase->transform(phase->zerocon(bt));
-    return VectorNode::scalar2vector(zero, length(), bt, bottom_type()->isa_vectmask() != nullptr);
+    return VectorNode::scalar2vector(zero, length(), bt,
+                                     bottom_type()->isa_vectmask() != nullptr);
   }
-  return VectorNode::Ideal(phase, can_reshape);
+
+  // Transformations for predicated IRs are not supported for now.
+  bool with_predicated = false;
+  if (is_predicated_vector() || in1->is_predicated_vector() ||
+      in2->is_predicated_vector()) {
+    with_predicated = true;
+  }
+
+  // For integer types:
+  // (XorV (VectorMaskCmp src1 src2 cond) (Replicate -1))
+  //    => (VectorMaskCmp src1 src2 ncond)
+  // (XorVMask (VectorMaskCmp src1 src2 cond) (MaskAll m1))
+  //    => (VectorMaskCmp src1 src2 ncond)
+  // cond can be eq, ne, le, ge, lt, gt, ule, uge, ult and ugt, ncond is the
+  // negative comparison of cond.
```

Comment:
```suggestion
  // cond can be eq, ne, le, ge, lt, gt, ule, uge, ult and ugt.
  // ncond is the negative comparison of cond.
```
I was getting lost in all the commas.

Reply:
Done.